### PR TITLE
nogo: Add a _base key to be a default config for all Analyzers.

### LINKED ISSUE #2513: nogo: include and exclude targets in configuration
### What version of rules_go are you using?
```
 http_archive(
        name = "io_bazel_rules_go",
        sha256 = "87f0fb9747854cb76a0a82430adccb6269f7d394237104a4523b51061c469171",
        urls = [
            "https://mirror.bazel.build/github.com/bazelbuild/rules_go/releases/download/v0.23.1/rules_go-v0.23.1.tar.gz",
            "https://github.com/bazelbuild/rules_go/releases/download/v0.23.1/rules_go-v0.23.1.tar.gz",
        ],
    )
```

### What version of gazelle are you using?

```
    http_archive(
        name = "bazel_gazelle",
        sha256 = "bfd86b3cbe855d6c16c6fce60d76bd51f5c8dbc9cfcaef7a2bb5c1aafd0710e8",
        urls = [
            "https://mirror.bazel.build/github.com/bazelbuild/bazel-gazelle/releases/download/v0.21.0/bazel-gazelle-v0.21.0.tar.gz",
            "https://github.com/bazelbuild/bazel-gazelle/releases/download/v0.21.0/bazel-gazelle-v0.21.0.tar.gz",
        ],
    )
```
### What version of Bazel are you using?
`bazel 3.1.0-homebrew`

### Does this issue reproduce with the latest releases of all the above?
Yes

### What operating system and processor architecture are you using?
Macos

### Any other potentially useful information about your toolchain?
I am using go modules. 

### What did you do?
`  go_register_toolchains(nogo = "@io_bazel_rules_go//:tools_nogo")`
I want to use nogo. I followed the instructions ( I know it mentions this) and it lines external packages which I have no control over. Is there a way I can limit it my code ? 

### What did you expect to see?
Lint only my code 

### What did you see instead?
Lints everything in the external. 
```
external/org_golang_x_tools/go/vcs/vcs.go:532:2: declaration of "metaImport" shadows declaration at line 599
```

### PR DESCRIPTION
**Pull request by DolceTriade**
This is useful for situations where you want to exclude/include the same directories for all the analyzers. Specifying a config for a particular analyzer will cause the _base config, if it exists, to be underlayed for that analyzer (ie, if the analyzer doesn't set an option, it will use the options specified by the _base config).

You can use it like so:
```
{
	"_base": {
		"exclude_files": {
			"external/": "third_party",
			"proto/": "generated protobuf"
		}
	},
	"override":
		"description": "oops, I actually don't want to exclude any files here...",
		"exclude_files": {}
	}
}
```
**What type of PR is this?**
Feature

**What does this PR do? Why is it needed?**
nogo: Add a _base key to be a default config for all Analyzers.

**Which issues(s) does this PR fix?**

Partially addresses #2513

**Other notes for review**
Now that we have agreed on the approach, this PR has been updated with a test and docs.
Thanks!

### CODE REVIEW

**Review by linzhp — COMMENTED** [review]
Thanks for the contribution! I think this is a good feature to have. One question: would it be more useful if we instead support a "_base" config, so an analyzer can override the "includes" but continue to use "excludes" from "_base", or vice versa?

**Comment by DolceTriade** [comment]
I thought about a `_base` option but I thought that a `_default` key was conceptually more clear. For example, for `_base` you need to support deleting entries in the `_base` config too. In most scenarios, if you are going to use the exact same settings for all the configs so no need to complicate the code by needing to merge the configs.

**Comment by linzhp** [comment]
An entry can be deleted by setting it to empty/null in the analyzer config. @jayconrod What do you think?

**Comment by jayconrod** [comment]
@DolceTriade Thanks for picking this up. I like this idea.

About `_base` vs `_default`: either seems fine. I like `_base` slightly better with the meaning that @linzhp suggested.

This is definitely an improvement, but I don't think it should be marked as fixing #2513. The configuration language still makes it hard to say which targets Nogo should apply to. Ideally it would accept Bazel labels with wildcards like:

```
{
  "printf": {
    "only_targets": {
      "//...": "Check targets in the main monorepo, not in external dependencies"
    },
    "exclude_targets": {
      "//vendor/...": "Ignore vendored packages"
    }
  }
}
```

(That's separate from this PR, and it's somewhat difficult since it means Nogo needs to parse, expand, and test Bazel label patterns. Let's just leave the issue open though).

**Comment by DolceTriade** [comment]
Got it. I'll change the implementation to the `_base` implementation.

```json
{
	"base": {
		"exclude_files": {
			"external/": "third_party",
			"proto/": "generated protobuf"
		}
	},
         "analyzer1": {
		"exclude_files": {
			"newpath/": "new path to scan",
			"proto/": null  # oops, turns out we actually want to scan this...
         }
}
```

**Comment by linzhp** [comment]
This may be easier to implement:

```json
{
	"base": {
		"exclude_files": {
			"external/": "third_party",
			"proto/": "generated protobuf"
		},
		"include_files": {
			"src": "all internal code",
		},

	},
         "analyzer1": {
		# include_files is inherited from "base", but exclude_files is overridden
		"exclude_files": {
			# "proto/" is not excluded here
			"newpath/": "new path to scan",
			"external/": "third_party",
		}
         },
         "analyzer2": {
		# include_files is inherited from "base", we don't want to exclude anything from analyzer2, so we set it to empty
		"exclude_files": {}
         },
}
```

**Comment by DolceTriade** [comment]
@linzhp Sounds good!

**Comment by DolceTriade** [comment]
Hello, I have updated the PR with the discussion. PTAL.

File: go/tools/builders/nogo_main.go [hunk 1]
```diff
@@ -401,12 +407,30 @@ func checkAnalysisResults(actions []*action, pkg *goPackage) string {
...
+			// the baseConfig config.
+		} else if baseConfig != nil {
+			if config.analyzerFlags == nil {
+				config.analyzerFlags = baseConfig.analyzerFlags
+			}
+			if config.onlyFiles == nil {
+				config.onlyFiles = baseConfig.onlyFiles
+			}
+			if config.excludeFiles == nil {
+				config.excludeFiles = baseConfig.excludeFiles
```
Comment (linzhp) [inline go/tools/builders/nogo_main.go:432]:
```go
for _, act := range actions {
	// ....
	var currentConfig config

	if baseConfig, ok := configs[nogoBaseConfigName]; ok {
		currentConfig = baseConfig
	}

	if actionConfig, ok := configs[act.a.Name]; ok {
		if actionConfig.analyzerFlags != nil {
			currentConfig.analyzerFlags = actionConfig.analyzerFlags
		}
		if actionConfig.onlyFiles != nil {
			currentConfig.onlyFiles = actionConfig.onlyFiles
		}
		if actionConfig.excludeFiles != nil {
			currentConfig.excludeFiles = actionConfig.excludeFiles
		}
	}

	if currentConfig.onlyFiles == nil && currentConfig.excludeFiles == nil {
		for _, diag := range act.diagnostics {
			diagnostics = append(diagnostics, entry{Diagnostic: diag, Analyzer: act.a})
		}
		continue
	}

	//...
}
```

I found this easier to read: for each action, starting from an empty config, using base config if it exists, overriding it with action-specific config if exists. If it's still empty at this point, emit diagnostic for all files and continue to next action.

Comment (DolceTriade) [reply go/tools/builders/nogo_main.go]: Ah, nice idea! Applied!

**Review by linzhp — APPROVED** [review]

**Pull request merged**